x-pack/metricbeat/module/prometheus/remote_write/data.go

```diff
@@ -0,0 +1,272 @@
+				upperBound := string(le.(model.LabelValue))
+
+				bucket, err := strconv.ParseFloat(upperBound, 64)
+				if err != nil {
+					continue
+				}
+				v := uint64(val)
+				b := &dto.Bucket{
+					CumulativeCount: &v,
+					UpperBound:      &bucket,
+				}
+				hist, ok := histograms[histKey]
+				if !ok {
+					hist = histogram{}
+				}
+				hist.buckets = append(hist.buckets, b)
+				hist.timestamp = metric.Timestamp.Time()
+				hist.labels = labelsClone
+				hist.metricName = name
+				histograms[histKey] = hist
+				continue
+			}
+			e.ModuleFields.Update(data)
+		}
+	}
+
+	// process histograms together
+	g.processPromHistograms(eventList, histograms)
+	return eventList
+}
+
+// rateCounterUint64 fills a counter value and optionally adds the rate if rate_counters is enabled
+func (g *remoteWriteTypedGenerator) rateCounterUint64(name string, labels common.MapStr, value uint64) common.MapStr {
+	d := common.MapStr{
+		"counter": value,
+	}
+
+	if g.rateCounters {
+		d["rate"], _ = g.counterCache.RateUint64(name+labels.String(), value)
+	}
+
+	return d
+}
+
+// rateCounterFloat64 fills a counter value and optionally adds the rate if rate_counters is enabled
+func (g *remoteWriteTypedGenerator) rateCounterFloat64(name string, labels common.MapStr, value float64) common.MapStr {
+	d := common.MapStr{
+		"counter": value,
+	}
+	if g.rateCounters {
+		d["rate"], _ = g.counterCache.RateFloat64(name+labels.String(), value)
+	}
+
+	return d
+}
+
+// processPromHistograms receives a group of Histograms and converts each one to ES histogram
+func (g *remoteWriteTypedGenerator) processPromHistograms(eventList map[string]mb.Event, histograms map[string]histogram) {
+	for _, histogram := range histograms {
+		labelsHash := histogram.labels.String()
+		if _, ok := eventList[labelsHash]; !ok {
+			eventList[labelsHash] = mb.Event{
+				ModuleFields: common.MapStr{},
+			}
+
+			// Add labels
+			if len(histogram.labels) > 0 {
+				eventList[labelsHash].ModuleFields["labels"] = histogram.labels
+			}
+		}
+
+		e := eventList[labelsHash]
+		e.Timestamp = histogram.timestamp
+
+		hist := dto.Histogram{
+			Bucket: histogram.buckets,
+		}
+		name := strings.TrimSuffix(histogram.metricName, "_bucket")
+		data := common.MapStr{
+			name: common.MapStr{
+				"histogram": collector.PromHistogramToES(g.counterCache, histogram.metricName, histogram.labels, &hist),
+			},
+		}
+		e.ModuleFields.Update(data)
+	}
+}
+
+// findMetricType evaluates the type of the metric by check the metricname format in order to handle it properly
+func (g *remoteWriteTypedGenerator) findMetricType(metricName string, labels common.MapStr) string {
+	leLabel := false
+	if _, ok := labels["le"]; ok {
+		leLabel = true
+	}
+	if strings.HasSuffix(metricName, "_total") || strings.HasSuffix(metricName, "_sum") ||
+		strings.HasSuffix(metricName, "_count") {
+		return counterType
+	} else if strings.Contains(metricName, "_bucket") && leLabel {
+		return histogramType
+	}
```

Comment:
What about having these as the defaults in the config parameters?

Reply:
hmm, tbh I'm not sure if we should expose these default values to the users. Users might accidentally remove/edit them and break the default case.

The only reason I would add them in the pattern list would be to give the option to the users to remove them in case they actually want it, however I don't think this would be the case since these patterns are quite straight forward.

lDo you see any case we could benefit of having these as config parameters?

Reply:
SGTM, let's go with how it is